docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,259 @@
+   > should mostly be a plumbing change, as long as we do it during a release
+   > in which we are not making other breaking changes to ABCI. However, the
+   > effort may be more or less depending on how gRPC integration works in the
+   > application's implementation language, and would have to be sure networks
+   > have plenty of time not only to make the change but to verify that it
+   > preserves the function of the network.
+   >
+   > What questions should we be asking node operators and application
+   > developers to understand the migration costs better?
+
+2. If we choose to keep only the socket protocol, we will need to follow up
+   with a more detailed design for extending and upgrading the protocol to fix
+   the existing performance and operational issues with the protocol.
+
+   Moreover, since the gRPC interface has been around for a long time we will
+   also need a deprecation plan for it.
+
+3. If we choose to keep both options, we will still need to do all the work of
+   (2), but the gRPC implementation should not require any immediate changes.
+
+
+## Alternatives Considered
+
+- **FFI**. Another approach we could take is to use a C-based FFI interface so
+  that applications written in other languages are linked directly with the
+  consensus node, an option currently only available for Go applications.
+
+  An FFI interface is possible for a lot of languages, but FFI support varies
+  widely in coverage and quality across languages and the points of friction
+  can be tricky to work around.  Moreover, it's much harder to add FFI support
+  to a language where it's missing after-the-fact for an application developer.
+
+  Although a basic FFI interface is not too difficult on the Go side, the C
+  shims for an FFI can get complicated if there's a lot of variability in the
+  runtime environment on the other end.
+
+  If we want to have one answer for non-Go applications, we are better off
+  picking an IPC-based solution (whether that's gRPC or an extension of our
+  custom socket protocol or something else).
+
+## Consequences
+
+- **Standardize on gRPC**
+
+   - ✅ Addresses existing performance and operational issues.
+   - ✅ Replaces custom code with a well-maintained widely-used library.
+   - ✅ Aligns with Cosmos SDK, which already uses gRPC extensively.
+   - ✅ Aligns with priv validator interface, for which the socket protocol is already deprecated for gRPC.
+   - ❓ Applications will be hard to implement in a language without gRPC support.
+   - ⛔ All users of the socket protocol have to migrate to gRPC, and we believe most current out-of-process applications use the socket protocol.
+
+- **Standardize on socket protocol**
+
+   - ✅ Less immediate impact for existing users (but see below).
+   - ✅ Simplifies ABCI API surface by removing gRPC.
+   - ❓ Users of the socket protocol will have a (smaller) migration.
+   - ❓ Potentially easier to implement for languages that do not have support.
```

Comment:
Why only potentially? Shouldn't it be much easier to implement than grpc ?

Reply:
Not necessarily that much easier, unless for some reason the language doesn't have a good HTTP client.